htlcswitch/quiescer.go

```diff
@@ -0,0 +1,123 @@
+package htlcswitch
+
+import (
+	"fmt"
+
+	"github.com/lightningnetwork/lnd/fn"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+// quiescer is a state machine that tracks progression through the quiescence
+// protocol.
+type quiescer struct {
+	// chanID marks what channel we are managing the state machine for. This
+	// is important because the quiescer is responsible for constructing the
+	// messages we send out and the ChannelID is a key field in that
+	// message.
+	chanID lnwire.ChannelID
+
+	// weOpened indicates whether we were the original opener of the
+	// channel. This is used to break ties when both sides of the channel
+	// send Stfu claiming to be the initiator.
+	weOpened bool
+
+	// localInit indicates whether our path through this state machine was
+	// initiated by our node. This can be true or false independently of
+	// remoteInit.
+	localInit bool
+
+	// remoteInit indicates whether we received Stfu from our peer where the
+	// message indicated that the remote node believes it was the initiator.
+	// This can be true or false independently of localInit.
+	remoteInit bool
+
+	// sent tracks whether or not we have emitted Stfu for sending.
+	sent bool
+
+	// received tracks whether or not we have received Stfu from our peer.
+	received bool
+}
+
+// recvStfu is called when we receive an Stfu message from the remote.
+func (q *quiescer) recvStfu(msg lnwire.Stfu) error {
+	if q.received {
+		return fmt.Errorf("stfu already received for channel %v",
+			q.chanID)
+	}
+
+	q.received = true
+	q.remoteInit = msg.Initiator
+
+	return nil
+}
+
+// sendStfu is called when we are ready to send an Stfu message. It returns the
+// Stfu message to be sent.
+func (q *quiescer) sendStfu() (fn.Option[lnwire.Stfu], error) {
+	if q.sent {
+		return fn.None[lnwire.Stfu](),
+			fmt.Errorf(
+				"stfu arleady sent for channel %s",
+				q.chanID.String(),
+			)
+	}
+
+	stfu := lnwire.Stfu{
+		ChanID:    q.chanID,
+		Initiator: q.localInit,
+	}
+
+	q.sent = true
+
+	return fn.Some(stfu), nil
+}
+
+// oweStfu returns true if we owe the other party an Stfu. We owe the remote an
+// Stfu when we have received but not yet sent an Stfu.
+func (q *quiescer) oweStfu() bool {
+	return q.received && !q.sent
+}
+
+// needStfu returns true if the remote owes us an Stfu. They owe us an Stfu when
+// we have sent but not yet received an Stfu.
+func (q *quiescer) needStfu() bool {
+	return q.sent && !q.received
+}
+
+// isQuiescent returns true if the state machine has been driven all the way to
+// completion. If this returns true, processes that depend on channel quiescence
+// may proceed.
+func (q *quiescer) isQuiescent() bool {
+	return q.sent && q.received
+}
+
+// isLocallyInitiatedFinal determines whether we are the initiator of quiescence
+// for the purposes of downstream protocols.
+func (q *quiescer) isLocallyInitiatedFinal() fn.Option[bool] {
```

Comment:
It can. I also don't like the name I went with here. There is an important "gotcha" with the protocol here, where both the local and remote party can initiate quiescence, and in that case the channel initiator gets the quiescence initiation role. The purpose of this function is not about whether or not we *tried* to initiate quiescence, but rather if we are the ultimate "initiator" wrt the quiescence *roles*.